File: client/modules/tyrs-ssh-transport/android/src/main/java/com/tyrshand/sshtransport/TyrsSSHTransportModule.kt
package com.tyrshand.sshtransport

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import org.json.JSONArray
import org.json.JSONObject
import sshtransport.Sshtransport

class TyrsSSHTransportModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("TyrsSSHTransport")

    AsyncFunction("openAppServer") { options: Map<String, Any?> ->
      parseObject(Sshtransport.openAppServer(
        options.string("profileId"), options.string("host"), options.long("port"),
        options.string("user"), options.string("privateKey"), options.optionalString("passphrase"),
        options.optionalString("expectedHostFingerprint"),
      ))
    }

    AsyncFunction("close") { profileId: String ->
      Sshtransport.close(profileId)
    }

    AsyncFunction("generateEd25519Key") {
      parseObject(Sshtransport.generateEd25519Key())
    }

    AsyncFunction("inspectPrivateKey") { privateKey: String, passphrase: String? ->
      parseObject(Sshtransport.inspectPrivateKey(privateKey, passphrase ?: ""))
    }

    AsyncFunction("probeHost") { options: Map<String, Any?> ->
      mapOf("fingerprint" to Sshtransport.probeHost(
        options.string("host"), options.long("port"), options.string("user"),
      ))
    }

    AsyncFunction("listDirectory") { options: Map<String, Any?> ->
      parseArray(Sshtransport.listDirectory(
        options.string("host"), options.long("port"), options.string("user"),
        options.string("privateKey"), options.optionalString("passphrase"),
        options.optionalString("expectedHostFingerprint"), options.string("path"),
      ))
    }

    AsyncFunction("uploadAttachment") { options: Map<String, Any?> ->
      parseObject(Sshtransport.uploadAttachment(
        options.string("host"), options.long("port"), options.string("user"),
        options.string("privateKey"), options.optionalString("passphrase"),
        options.optionalString("expectedHostFingerprint"), options.string("localPath"),
        options.string("filename"), options.optionalString("mimeType"),
      ))
    }
  }
}

private fun Map<String, Any?>.string(name: String): String =
  this[name] as? String ?: throw IllegalArgumentException("缺少 $name")

private fun Map<String, Any?>.optionalString(name: String): String = this[name] as? String ?: ""

private fun Map<String, Any?>.long(name: String): Long =
  (this[name] as? Number)?.toLong() ?: throw IllegalArgumentException("缺少 $name")

private fun parseObject(encoded: String): Map<String, Any?> = jsonObject(JSONObject(encoded))

private fun parseArray(encoded: String): List<Any?> = jsonArray(JSONArray(encoded))

private fun jsonObject(value: JSONObject): Map<String, Any?> = value.keys().asSequence()
  .associateWith { key -> jsonValue(value.get(key)) }

private fun jsonArray(value: JSONArray): List<Any?> =
  (0 until value.length()).map { index -> jsonValue(value.get(index)) }

private fun jsonValue(value: Any?): Any? = when (value) {
  JSONObject.NULL -> null
  is JSONObject -> jsonObject(value)
  is JSONArray -> jsonArray(value)
  else -> value
}

File: client/modules/tyrs-ssh-transport/ios/TyrsSSHTransportModule.swift
import ExpoModulesCore
import Sshtransport

public final class TyrsSSHTransportModule: Module {
  public func definition() -> ModuleDefinition {
    Name("TyrsSSHTransport")

    AsyncFunction("openAppServer") { (options: [String: Any]) -> [String: Any] in
      try parseObject(goString { error in try SshtransportOpenAppServer(
        string(options, "profileId"), string(options, "host"), integer(options, "port"),
        string(options, "user"), string(options, "privateKey"),
        optionalString(options, "passphrase"),
        optionalString(options, "expectedHostFingerprint"), error
      ) })
    }

    AsyncFunction("close") { (profileId: String) in
      SshtransportClose(profileId)
    }

    AsyncFunction("inspectRuntime") { (options: [String: Any]) -> [String: Any] in
      try parseObject(goString { error in try SshtransportInspectRuntime(
        string(options, "host"), integer(options, "port"), string(options, "user"),
        string(options, "privateKey"), optionalString(options, "passphrase"),
        string(options, "expectedHostFingerprint"), error
      ) })
    }

    AsyncFunction("generateEd25519Key") { () -> [String: Any] in
      try parseObject(goString { SshtransportGenerateEd25519Key($0) })
    }

    AsyncFunction("inspectPrivateKey") { (privateKey: String, passphrase: String?) -> [String: Any] in
      try parseObject(goString { SshtransportInspectPrivateKey(privateKey, passphrase ?? "", $0) })
    }

    AsyncFunction("probeHost") { (options: [String: Any]) -> [String: Any] in
      ["fingerprint": try goString { error in try SshtransportProbeHost(
        string(options, "host"), integer(options, "port"), string(options, "user"), error
      ) }]
    }

    AsyncFunction("listDirectory") { (options: [String: Any]) -> [Any] in
      try parseArray(goString { error in try SshtransportListDirectory(
        string(options, "host"), integer(options, "port"), string(options, "user"),
        string(options, "privateKey"), optionalString(options, "passphrase"),
        optionalString(options, "expectedHostFingerprint"), string(options, "path"), error
      ) })
    }

    AsyncFunction("uploadAttachment") { (options: [String: Any]) -> [String: Any] in
      try parseObject(goString { error in try SshtransportUploadAttachment(
        string(options, "host"), integer(options, "port"), string(options, "user"),
        string(options, "privateKey"), optionalString(options, "passphrase"),
        optionalString(options, "expectedHostFingerprint"), string(options, "localPath"),
        string(options, "filename"), optionalString(options, "mimeType"), error
      ) })
    }

    AsyncFunction("downloadFile") { (options: [String: Any]) -> [String: Any] in
      try parseObject(goString { error in try SshtransportDownloadFile(
        string(options, "host"), integer(options, "port"), string(options, "user"),
        string(options, "privateKey"), optionalString(options, "passphrase"),
        optionalString(options, "expectedHostFingerprint"), string(options, "remotePath"),
        string(options, "localPath"), error
      ) })
    }
  }
}

private enum SSHTransportError: Error {
  case missingOption(String)
  case invalidJSON
}

private func string(_ options: [String: Any], _ name: String) throws -> String {
  guard let value = options[name] as? String else {
    throw SSHTransportError.missingOption(name)
  }
  return value
}

private func optionalString(_ options: [String: Any], _ name: String) -> String {
  options[name] as? String ?? ""
}

private func integer(_ options: [String: Any], _ name: String) throws -> Int {
  guard let value = options[name] as? NSNumber else {
    throw SSHTransportError.missingOption(name)
  }
  return value.intValue
}

// gomobile 返回非空 NSString，Swift 不会自动把 NSError 指针转换为 throws。
private func goString(_ call: (NSErrorPointer) throws -> String) throws -> String {
  var error: NSError?
  let value = try call(&error)
  if let error { throw error }
  return value
}

private func parseObject(_ encoded: String) throws -> [String: Any] {
  guard let data = encoded.data(using: .utf8),
        let value = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
    throw SSHTransportError.invalidJSON
  }
  return value
}

private func parseArray(_ encoded: String) throws -> [Any] {
  guard let data = encoded.data(using: .utf8),
        let value = try JSONSerialization.jsonObject(with: data) as? [Any] else {
    throw SSHTransportError.invalidJSON
  }
  return value
}
